### PR TITLE
Fish compatibility

### PR DESCRIPTION
**Pull request by lordlycastle**
It should work now with fish. Though I couldn't figure out how to add the command file to brew, right now it doesn't recognises the `brew command command-not-found-init-fish`. Also there isn't a test for it, though I don't think there is a need.

### CODE REVIEW

**Comment by bfontaine** [comment]
Awesome :smile:

File: README.md [hunk 1]
```diff
@@ -2,11 +2,10 @@
...
 [![Build Status](https://travis-ci.org/Homebrew/homebrew-command-not-found.svg?branch=master)](https://travis-ci.org/Homebrew/homebrew-command-not-found)
 
-This project tries to reproduce Ubuntu’s `command-not-found` for Homebrew users
-on OSX.
+This project tries to reproduce Ubuntu’s `command-not-found` for Homebrew users on OSX.
+
+On Ubuntu, when you try to use a command that doesn’t exist locally but is available through a package, your shell will suggest you a command to install it.
 
-On Ubuntu, when you try to use a command that doesn’t exist locally but is
-available through a package, Bash will suggest you a command to install it.
```
Comment (bfontaine) [inline README.md]: Please leave this text as is, lines should fit under 80 characters.

File: README.md [hunk 2]
```diff
@@ -23,13 +22,18 @@ The program 'when' is currently not installed. You can install it by typing:
...
 ## Install
 
-First, tap this repository:
-
-    brew tap homebrew/command-not-found
-
-Then add following line to your `.bashrc`(bash) or `.zshrc`(zsh):
+First, tap this repository: `brew tap homebrew/command-not-found`
 
+* **Bash and Zsh**: Add following line to your `.bashrc`(bash) or `.zshrc`(zsh):
```
Comment (bfontaine) [inline README.md]: > Add *the* following line

Also, please add a space between `` `.bashrc` `` and `(bash)` (same thing for zsh).

File: README.md [hunk 3]
```diff
@@ -43,23 +47,14 @@ upgrade to 0.2.0 just remove it and follow the install instructions above.
...
+
+For Bash and Zsh, the `handler.sh` script defines a function `command_not_found_handle` and `command_not_found_handler`, respectively. Which is used when you try a command that doesn’t exist. The function calls `brew which-formula` on your command, and if it finds a match it’ll print it to you. If not, you’ll get an error as expected.
+
+For Fish, the `handler.fish` defines a function `command-not-found-handler` with is used by Fish when a command doesn't exist. The function is copied in your functions directory. Fish recognises the function, and uses this instead of it's default handler. The function utilises `brew which-formula` to find, and prints the relevant information about the command; on the case where it finds nothing, the function run the default handler (`__fish_default_command_not_found_handler`).
+
+Over 4300 formulae have been imported in the database, representing more than 19900 different commands (99% of the main Homebrew repo + 69% of all official taps).
```
Comment (bfontaine) [inline README.md]: Please keep lines below 80 characters, as they are now.

File: cmd/brew-command-not-found-init-fish.rb [hunk 4]
```diff
@@ -0,0 +1,7 @@
+
+require "fileutils"
+
+file = File.expand_path("#{File.dirname(__FILE__)}/../handler.fish")
+destination = ARGV[0] ? ARGV[0]+"command-not-found.fish" : File.expand_path("~/.config/fish/functions/command-not-found.fish")
+
+FileUtils.cp(file, destination)
```
Comment (bfontaine) [inline cmd/brew-command-not-found-init-fish.rb]: I’d prefer if we could have only one command, we can use an option e.g.: `brew command-not-found-init --fish`.

File: cmd/brew-command-not-found-init-fish.rb [hunk 5]
```diff
@@ -0,0 +1,7 @@
+
+require "fileutils"
+
+file = File.expand_path("#{File.dirname(__FILE__)}/../handler.fish")
+destination = ARGV[0] ? ARGV[0]+"command-not-found.fish" : File.expand_path("~/.config/fish/functions/command-not-found.fish")
```
Comment (bfontaine) [inline cmd/brew-command-not-found-init-fish.rb]: Use `ARGV.empty?` to test if any argument was passed. Also, `ARGV[0]+"command-not-found.fish"` won’t work if `ARGV[0]` doesn’t end with a slash. Use `File.join(ARGV[0], "command-not-found.fish")` to be sure.

File: cmd/brew-command-not-found-init-fish.rb (hunk 4 above)
Comment (bfontaine) [inline cmd/brew-command-not-found-init-fish.rb]: Can’t we load the file without copying it into the user’s directory?

File: README.md [hunk 6]
```diff
@@ -43,23 +47,14 @@ upgrade to 0.2.0 just remove it and follow the install instructions above.
...
+
+For Bash and Zsh, the `handler.sh` script defines a function `command_not_found_handle` and `command_not_found_handler`, respectively. Which is used when you try a command that doesn’t exist. The function calls `brew which-formula` on your command, and if it finds a match it’ll print it to you. If not, you’ll get an error as expected.
+
+For Fish, the `handler.fish` defines a function `command-not-found-handler` with is used by Fish when a command doesn't exist. The function is copied in your functions directory. Fish recognises the function, and uses this instead of it's default handler. The function utilises `brew which-formula` to find, and prints the relevant information about the command; on the case where it finds nothing, the function run the default handler (`__fish_default_command_not_found_handler`).
```
Comment (bfontaine) [inline README.md]: > instead of *its* default handler

 
> to find and print

 
> the function *runs*

**Comment by lordlycastle** [comment]
Thanks for all the feedback. I tried to load the function without coping, but couldn't get it work (I'll think about it more). I wouldn't merge it just yet, there was a significant lag when starting Terminal. Perhaps there is better way...

**Comment by lordlycastle** [comment]
I don't think eval will work in fish, because it'll run all the command in a block and the function will be erased when block goes out of scope. Someone with more fish experience could help more. 

As for the lag it's about 0.3-0.4 seconds at the start which is due to having to copy the file. What if command-not-found was a brew formula too, so you could install it after tapping. This way the function file will have to copied only once, or with every update.

**Comment by bfontaine** [comment]
What if, instead of copying the file, we asked users to make a symlink to the handler? They would do it once and would always get all updates since it would be a symlink to the versionned version.

We could even do it for them with a command like `brew command-not-found-init --fish`.

**Comment by lordlycastle** [comment]
That might work, do you know the ruby function for that? I'm not familiar with ruby so what would be the best way to create an alias.

**Comment by lordlycastle** [comment]
I think 7a9e8dc is ready to be merged. Unless you have any more feedback?

**Comment by bfontaine** [comment]
@lordlycastle Sorry I’m currently busy preparing a trip, I won’t have access to my computer between tomorrow and next Monday. I’ll look at it when I’ll be back :)

**Comment by bfontaine** [comment]
@lordlycastle Could you please rebase your branch on top of master?

**Comment by lordlycastle** [comment]
@bfontaine I'm sorry but I don't know how to do that, I'm still very new to Git, any guidance would be much appreciated.

**Comment by bfontaine** [comment]
@lordlycastle I temporary reverted the commit which causes the conflict because it’s hard to fix it. So you don’t have to do anything on your side.

File: README.md [hunk 7]
```diff
@@ -6,10 +6,11 @@ This project tries to reproduce Ubuntu’s `command-not-found` for Homebrew user
 on OSX.
 
 On Ubuntu, when you try to use a command that doesn’t exist locally but is
-available through a package, Bash will suggest you a command to install it.
+available through a package, your shell will suggest you a command to install it.
+
```
Comment (bfontaine) [inline README.md]: Please don’t change this, it’s still true because it’s talking about Ubuntu, not OS X.

File: README.md [hunk 8]
```diff
@@ -6,10 +6,11 @@ This project tries to reproduce Ubuntu’s `command-not-found` for Homebrew user
...
 Using this script, you can replicate this feature on OSX:
 
-```
+```bash
```
Comment (bfontaine) [inline README.md]: Please don’t add this, the code snippet is technically not in Bash (e.g. `$ when` include a prompt).

File: README.md [hunk 9]
```diff
@@ -23,13 +24,24 @@ The program 'when' is currently not installed. You can install it by typing:
 
 ## Install
 
-First, tap this repository:
-
-    brew tap homebrew/command-not-found
-
-Then add following line to your `.bashrc`(bash) or `.zshrc`(zsh):
+First, tap this repository: `brew tap homebrew/command-not-found`
```
Comment (bfontaine) [inline README.md]: Please keep the code block here.

File: README.md [hunk 10]
```diff
@@ -23,13 +24,24 @@ The program 'when' is currently not installed. You can install it by typing:
...
 
+* **Bash and Zsh**: Add following line to your `.bashrc` (bash) or `.zshrc` (zsh):
```
Comment (bfontaine) [inline README.md]: > Add **the** following line

File: README.md [hunk 11]
```diff
@@ -43,7 +55,7 @@ upgrade to 0.2.0 just remove it and follow the install instructions above.
 
 ## Support
 
-This tool supports Bash (version 4 and higher) and Zsh.
+This tool supports Bash (version 4 and higher), Zsh, and Fish (2.2.0)
```
Comment (bfontaine) [inline README.md]: Please add a dot at the end of the sentence.

File: cmd/brew-command-not-found-init.rb [hunk 12]
```diff
@@ -1,3 +1,11 @@
 #!/usr/bin/env ruby
+require "fileutils"
 
-puts File.read(File.expand_path "#{File.dirname(__FILE__)}/../handler.sh")
+if (!ARGV.empty? and ARGV[0] == "--fish")
```
Comment (bfontaine) [inline cmd/brew-command-not-found-init.rb]: use `if ARGV.first == "--fish"`

File: cmd/brew-command-not-found-init.rb [hunk 13]
```diff
@@ -1,3 +1,11 @@
...
+	file = File.expand_path("#{File.dirname(__FILE__)}/../handler.fish")
+	destination = (ARGV[1] != nil) ? File.expand_path(File.join(ARGV[1], "command-not-found.fish")) : File.expand_path("~/.config/fish/functions/command-not-found.fish")
```
Comment (bfontaine) [inline cmd/brew-command-not-found-init.rb]: This line is too long, please use a multiline condition, e.g.:

```
if …
  destination = …
else
  destination = …
end
```

File: cmd/brew-command-not-found-init.rb [hunk 14]
```diff
@@ -1,3 +1,11 @@
 #!/usr/bin/env ruby
+require "fileutils"
```
Comment (bfontaine) [inline cmd/brew-command-not-found-init.rb]: You can move this in the `if` block:

```rb
if ARGV.first == "--fish"
  require "fileutils"
  # …
```

**Comment by lordlycastle** [comment]
@bfontaine Got them. Thanks for the feedback :+1:.

**Comment by xu-cheng** [comment]
**NOTE**: I'm not a fish user, all of below comes from google and experience.

It seems that you can avoid the file manipulation complete. If you do this:

* Make `brew command-not-found-init` print the shell script based on current shell.
* Add below codes to your fish config:
```fish
status --is-interactive; and brew command command-not-found-init ^ /dev/null; and . (brew command-not-found-init | psub)
```

File: handler.fish [hunk 15]
```diff
@@ -0,0 +1,10 @@
+function command-not-found
+	set -l cmd "$argv[2]"
+	set -l txt (brew which-formula --explain $cmd ^/dev/null)
+	
+	if test -z "$txt"
```
Comment (xu-cheng) [inline handler.fish]: We also need to check whether it's interactive shell.

File: cmd/brew-command-not-found-init.rb [hunk 16]
```diff
@@ -1,3 +1,16 @@
 #!/usr/bin/env ruby
 
-puts File.read(File.expand_path "#{File.dirname(__FILE__)}/../handler.sh")
+if (ARGV.first == "--fish")
+	require "fileutils"
+	file = File.expand_path("#{File.dirname(__FILE__)}/../handler.fish")
+	
+	if (ARGV[1] != nil)
```
Comment (bfontaine) [inline cmd/brew-command-not-found-init.rb]: Please remove the parentheses here, we don’t need them.

**Comment by xu-cheng** [comment]
> I don't think eval will work in fish, because it'll run all the command in a block and the function will be erased when block goes out of scope. Someone with more fish experience could help more.

Test shows `eval` works fine for me.

**Pull request closed without merging**